### PR TITLE
feat: AWS codepipeline

### LINKED ISSUE #627: AWS::CodePipeline::Pipeline

### PR DESCRIPTION
**Pull request by zagronitay**

### CODE REVIEW

File: resources/provider/migrations/postgres/24_v0.10.16.up.sql [hunk 1]
```diff
@@ -1,6 +1,57 @@
 -- Autogenerated by migration tool on 2022-04-04 17:12:37
 -- CHANGEME: Verify or edit this file before proceeding
 
+-- Resource: codepipeline.pipelines
+CREATE TABLE IF NOT EXISTS "aws_codepipeline_pipelines" (
+                                                            "cq_id" uuid NOT NULL,
```
Comment (irmatov) [inline resources/provider/migrations/postgres/24_v0.10.16.up.sql:6]: nit: weird formatting. Did you reformat it manually like that from the output of migration tool?

File: resources/services/codepipeline/pipelines.go [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+	svc := c.Services().CodePipeline
+	config := codepipeline.ListPipelinesInput{}
+	for {
+		response, err := svc.ListPipelines(ctx, &config, func(options *codepipeline.Options) {
+			options.Region = c.Region
+		})
+		if err != nil {
+			return diag.WrapError(err)
+		}
+		if len(response.Pipelines) == 0 {
```
Comment (irmatov) [inline resources/services/codepipeline/pipelines.go:230]: nit: not really necessary

File: resources/services/codepipeline/pipelines.go [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+	res <- r.Pipeline.Stages
+	return nil
+}
+func resolvePipelineStagesBlockers(ctx context.Context, meta schema.ClientMeta, resource *schema.Resource, c schema.Column) error {
+	r := resource.Item.(types.StageDeclaration)
+	data, err := json.Marshal(r.Blockers)
+	if err != nil {
+		return diag.WrapError(err)
+	}
+	return resource.Set(c.Name, data)
```
Comment (irmatov) [inline resources/services/codepipeline/pipelines.go:265]: `return diag.WrapError(resource.Set(c.Name, data))`

File: resources/services/codepipeline/pipelines.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+	res <- r.Actions
+	return nil
+}
+func resolvePipelineStageActionsInputArtifacts(ctx context.Context, meta schema.ClientMeta, resource *schema.Resource, c schema.Column) error {
+	r := resource.Item.(types.ActionDeclaration)
+	artifacts := make([]*string, 0)
+	for _, a := range r.InputArtifacts {
+		artifacts = append(artifacts, a.Name)
+	}
+	return resource.Set(c.Name, artifacts)
```
Comment (irmatov) [inline resources/services/codepipeline/pipelines.go:278]: `return diag.WrapError(resource.Set(c.Name, data))`

File: resources/services/codepipeline/pipelines.go [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+}
+func resolvePipelineStageActionsOutputArtifacts(ctx context.Context, meta schema.ClientMeta, resource *schema.Resource, c schema.Column) error {
+	r := resource.Item.(types.ActionDeclaration)
+	artifacts := make([]*string, 0)
+	for _, a := range r.OutputArtifacts {
+		artifacts = append(artifacts, a.Name)
+	}
+	return resource.Set(c.Name, artifacts)
```
Comment (irmatov) [inline resources/services/codepipeline/pipelines.go:286]: `return diag.WrapError(resource.Set(c.Name, data))`

File: resources/services/codepipeline/pipelines.go [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+			},
+			{
+				Name:        "artifact_store_encryption_key_id",
+				Description: "The ID used to identify the key",
+				Type:        schema.TypeString,
+				Resolver:    schema.PathResolver("Pipeline.ArtifactStore.EncryptionKey.Id"),
+			},
+			{
+				Name:        "artifact_store_encryption_key_type",
+				Description: "The type of encryption key, such as an AWS Key Management Service (AWS KMS) key. When creating or updating a pipeline, the value must be set to 'KMS'.  This member is required.",
```
Comment (irmatov) [inline resources/services/codepipeline/pipelines.go:88]: nit: comment can be trimmed after the first sentence.

**Review by irmatov — APPROVED** [review]
LGTM, although there are some minor suggestions

File: resources/provider/migrations/postgres/24_v0.10.16.up.sql (hunk 1 above)
Comment (zagronitay) [reply resources/provider/migrations/postgres/24_v0.10.16.up.sql:6]: migration tool but I will check again

Comment (disq) [reply resources/provider/migrations/postgres/24_v0.10.16.up.sql:6]: Looks like GoLand's weird SQL formatting to me

Comment (zagronitay) [reply resources/provider/migrations/postgres/24_v0.10.16.up.sql]: Yes, probably
Fixed

File: client/services.go [hunk 7]
```diff
@@ -598,3 +599,9 @@ type WorkspacesClient interface {
 	DescribeWorkspaces(ctx context.Context, params *workspaces.DescribeWorkspacesInput, optFns ...func(*workspaces.Options)) (*workspaces.DescribeWorkspacesOutput, error)
 	DescribeWorkspaceDirectories(ctx context.Context, params *workspaces.DescribeWorkspaceDirectoriesInput, optFns ...func(*workspaces.Options)) (*workspaces.DescribeWorkspaceDirectoriesOutput, error)
 }
+
+//go:generate mockgen -package=mocks -destination=./mocks/mock_codepipeline.go . CodePipelineClient
+type CodePipelineClient interface {
```
Comment (roneli) [inline client/services.go:604]: would maybe add https://docs.aws.amazon.com/codepipeline/latest/APIReference/API_ListWebhooks.html API, would let @bbernays decide if we should do this in this PR or another issue.

Comment (roneli) [inline client/services.go:604]: missing https://docs.aws.amazon.com/codepipeline/latest/APIReference/API_ListTagsForResource.html to get tags for all pipelines

File: resources/provider/migrations/postgres/24_v0.10.16.up.sql [hunk 8]
```diff
@@ -1,6 +1,57 @@
...
+      FOREIGN KEY (pipeline_cq_id) REFERENCES aws_codepipeline_pipelines(cq_id) ON DELETE CASCADE
+);
+CREATE TABLE IF NOT EXISTS "aws_codepipeline_pipeline_stage_actions" (
+     "cq_id" uuid NOT NULL,
+     "cq_meta" jsonb,
+     "pipeline_stage_cq_id" uuid,
+     "action_type_id_category" text,
+     "action_type_id_owner" text,
+     "action_type_id_provider" text,
+     "action_type_id_version" text,
```
Comment (roneli) [inline resources/provider/migrations/postgres/24_v0.10.16.up.sql:41]: I don't think there is need for action_type_id prefix.

File: resources/services/codepipeline/pipelines.go [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+		if err != nil {
+			return diag.WrapError(err)
+		}
+		for i := range response.Pipelines {
+			response, err := svc.GetPipeline(ctx, &codepipeline.GetPipelineInput{Name: response.Pipelines[i].Name}, func(o *codepipeline.Options) {
+				o.Region = c.Region
+			})
+			if err != nil {
+				if c.IsNotFoundError(err) {
+					continue
```
Comment (roneli) [inline resources/services/codepipeline/pipelines.go:241]: add a debug log line here.

**Review by roneli — CHANGES_REQUESTED** [review]
Looks good overall, added some review comments

File: client/services.go (hunk 7 above)
Comment (bbernays) [reply client/services.go:604]: We can do it in another PR. I have opened an issue to track [that work:](https://github.com/cloudquery/cq-provider-aws/issues/658)

File: resources/services/codepipeline/pipelines.go [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+				Resolver:    client.ResolveAWSAccount,
+			},
+			{
+				Name:        "region",
+				Description: "The AWS Region of the resource.",
+				Type:        schema.TypeString,
+				Resolver:    client.ResolveAWSRegion,
+			},
+			{
+				Name:     "tags",
```
Comment (roneli) [inline resources/services/codepipeline/pipelines.go:39]: missing description

File: resources/services/codepipeline/pipelines.go [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+						Resolver:    schema.ParentIdResolver,
+					},
+					{
+						Name:        "name",
+						Description: "The name of the stage.  This member is required.",
+						Type:        schema.TypeString,
+					},
+					{
+						Name:          "blockers",
+						Description:   "Reserved for future use.",
```
Comment (roneli) [inline resources/services/codepipeline/pipelines.go:130]: maybe modify this description?

**Review by roneli — COMMENTED** [review]
LGTM, added had two comments regarding descriptions.

File: resources/provider/migrations/postgres/24_v0.10.16.up.sql (hunk 1 above)
Comment (disq) [reply resources/provider/migrations/postgres/24_v0.10.16.up.sql:6]: Looks like it's back?

File: resources/provider/migrations/postgres/24_v0.10.16.up.sql [hunk 12]
```diff
@@ -1,6 +1,58 @@
...
+                                                            "artifact_store_type" text,
+                                                            "artifact_store_encryption_key_id" text,
+                                                            "artifact_store_encryption_key_type" text,
+                                                            "artifact_stores" jsonb,
+                                                            "version" integer,
+                                                            CONSTRAINT aws_codepipeline_pipelines_pk PRIMARY KEY(arn),
+                                                            UNIQUE(cq_id)
+);
+CREATE TABLE IF NOT EXISTS "aws_codepipeline_pipeline_stages" (
+                                                                  "cq_id" uuid NOT NULL,
```
Comment (disq) [inline resources/provider/migrations/postgres/24_v0.10.16.up.sql:26]: I think you're missing the stage ordering information here?

Comment (disq) [reply resources/provider/migrations/postgres/24_v0.10.16.up.sql:26]: I see there's `run_order` in stage_actions, but don't the stages themselves have order within a pipeline? I don't know, just asking. Terraform `stage` blocks also don't seem to have explicit ordering apart from being separate blocks.

Comment (bbernays) [reply resources/provider/migrations/postgres/24_v0.10.16.up.sql:26]: Great call out. I think you are right that the stage ordering is implicit to how the blocks are ordered. That is also how it is in Cloudformation. Looking at the [CLI](https://docs.aws.amazon.com/cli/latest/reference/codepipeline/update-pipeline.html) there is no way to update a single stage, you have to specify all of the stages. As such we should create a column to specify the index/order of each stage

File: resources/services/codepipeline/pipelines.go [hunk 13]
```diff
@@ -120,6 +121,11 @@ func Pipelines() *schema.Table {
 						Type:        schema.TypeUUID,
 						Resolver:    schema.ParentIdResolver,
 					},
+					{
+						Name:        "order",
```
Comment (disq) [inline resources/services/codepipeline/pipelines.go:125]: reserved word, could create problems or inconvenience

**Review by disq — APPROVED** [review]

**Review by bbernays — APPROVED** [review]

**Review by roneli — APPROVED** [review]
LGTM

**Pull request merged**